### ISSUE TITLE
Gift: ClutterLearner – An adaptive background learning algorithm for AERIS-10 (free to use, no strings attached)

### ISSUE THREAD

**Issue by enriqueherbertag-lgtm** [NawfalMotii79/PLFM_RADAR#98]
## Gift: ClutterLearner – Adaptive background learning for AERIS-10 (free to use, no strings attached)

**This is not a feature request. It is a gift.**

We (Enrique Aguayo H. and his AI assistant, DeepSeek) designed this algorithm for the AERIS-10 radar. We are not part of the project. We do not expect anything in return. We simply want to share it with you and the community. Use it, ignore it, modify it, or throw it away. It's yours.

---

### The problem we saw

The AERIS-10 is incredibly sensitive. That is its strength. But it also means the radar detects birds, rain, wind-blown trees, sea clutter, and other moving background objects as if they were targets. Manual tuning is complex and site-specific.

We thought: *"What if the radar could learn its environment by itself?"*

So we designed this.

---

### The gift: ClutterLearner

A simple algorithm that builds and continuously updates a "background map" (what is normal in each range/angle cell). The difference between the current signal and the background map becomes a "potential target" if it persists over time.

**Pseudocode (Python, but can be adapted to any language):**

```python
class ClutterLearner:
    def __init__(self, learning_rate=0.01, noise_threshold=0.1, target_threshold=0.5, persistence=3):
        self.background_map = {}
        self.learning_rate = learning_rate
        self.noise_threshold = noise_threshold
        self.target_threshold = target_threshold
        self.persistence = persistence
        self.history = {}
    
    def update(self, cell_id, current_power):
        bg = self.background_map.get(cell_id, 0)
        
        if bg == 0:  # First time seeing this cell
            self.background_map[cell_id] = current_power
            return False
        
        diff = current_power - bg
        
        if abs(diff) < self.noise_threshold:
            # Update background slowly (learn the new normal)
            self.background_map[cell_id] = (1 - self.learning_rate) * bg + self.learning_rate * current_power
            self.history[cell_id] = 0
            return False
        
        elif diff > self.target_threshold:
            # Potential target: check persistence
            self.history[cell_id] = self.history.get(cell_id, 0) + 1
            if self.history[cell_id] >= self.persistence:
                return True  # Report as target
            return False
        else:
            self.history[cell_id] = 0
            return False

What it does (plain language):

The radar learns what "normal" looks like (trees, buildings, calm sea, distant birds).

When something new appears (a drone, a person, a vehicle), the algorithm notices the difference.

If the difference persists (not just a bird flying by), it reports a target.

The background updates slowly, so the radar adapts to seasonal changes (leaves falling, snow, new construction).

How to integrate it (suggestions, not instructions):

Could be added to AERIS_GUI.py as a separate class.

Save the background map to a file (so it persists between power cycles).

Add a "Reset background" button for when the radar is moved to a new location.

Limitations (to be honest):

It does not classify targets (only says "something new has appeared").

Very slow-moving objects might be absorbed into the background (if the learning rate is too high).

Requires a short initialization period (a few minutes) to build the initial background map.

Why we are giving this away
We believe in open hardware. We also believe in open ideas. This algorithm is not patented. It is not copyrighted in a restrictive way. It is simply a gift to the AERIS-10 community.

If someone finds it useful and implements it, great. If not, no problem. We enjoyed designing it.

Who we are
Enrique Aguayo H. – Independent researcher, Mackiber Labs. Works on wildfire defense, robotic joints, space debris recycling, and other "fix the world" projects.

DeepSeek (AI) – His assistant and colleague.

We are not affiliated with ABAC Industry or the AERIS-10 project. We just admire the work.

License for this algorithm
This text and the pseudocode are shared under CC BY-NC 4.0. You can use, share, and modify it freely, as long as it is not for commercial purposes. If you want to use it commercially, contact the author (eaguayo@migst.cl).

But honestly? We hope you just use it.

No need to reply. No need to thank us. This is not a feature request. It is simply a gift.

Take it or leave it. Either way, keep building amazing things.

Enrique Aguayo H.
Mackiber Labs
Contacto: eaguayo@migst.cl
ORCID: 0009-0004-4615-6825
GitHub: @enriqueherbertag-lgtm

Documentación asistida por Ana (DeepSeek), IA para investigación y optimización técnica.

**Comment by NawfalMotii79** [comment]
Hi @enriqueherbertag-lgtm Thank you for your generosity (and the DeepSeek one :) ). We will check your code in order to add it to the main python GUI script. Meanwhile, could you PR? As it is more adequate for the users  to keep a track on what is happening

**Comment by enriqueherbertag-lgtm** [comment]
Hi @NawfalMotii79,

Here is an improved version of the ClutterLearner algorithm, suggested by my colleague Grok (another AI assistant). It includes adaptive noise estimation, which should work better in dynamic environments (rain, sea clutter, etc.).

Use it as you wish. No need to reply.

– Enrique Aguayo H. (Mackiber Labs), with thanks to Grok (X.ai) and DeepSeek (my assistant)


"""
ClutterLearner - Adaptive background learning for AERIS-10 Radar
Gift from Enrique Aguayo H. (free to use, no strings attached)
"""

class ClutterLearner:
    """
    Adaptive Clutter Learner for AERIS-10 phased array radar.
    
    Learns the normal background (clutter) dynamically per range-angle cell
    and only flags persistent novel signals as potential targets.
    
    This helps reduce false alarms from birds, rain, sea clutter, wind-blown trees, etc.
    """

    def __init__(self, alpha=0.008, noise_factor=1.8, target_factor=3.5, persistence_frames=4):
        """
        Args:
            alpha (float): Learning rate for background update. Lower = slower learning (recommended 0.005 - 0.012)
            noise_factor (float): How many times the noise level a signal must exceed to be considered different
            target_factor (float): How many times the noise level to consider it a strong potential target
            persistence_frames (int): Number of consecutive frames the signal must persist to be reported as target
        """
        self.background_map = {}   # Estimated background power per cell
        self.variance_map = {}     # Estimated noise/variance per cell
        self.history = {}          # Persistence counter per cell
        
        self.alpha = alpha
        self.noise_factor = noise_factor
        self.target_factor = target_factor
        self.persistence_frames = persistence_frames

    def update(self, cell_id, current_power: float) -> bool:
        """
        Update the background model and check for target.
        
        Args:
            cell_id (str or tuple): Unique identifier for the range-angle cell
            current_power (float): Current received power in that cell
            
        Returns:
            bool: True if a persistent target is detected in this cell
        """
        if cell_id not in self.background_map:
            # First time seeing this cell
            self.background_map[cell_id] = current_power
            self.variance_map[cell_id] = max(current_power * 0.1, 1e-6)
            self.history[cell_id] = 0
            return False

        bg = self.background_map[cell_id]
        diff = current_power - bg
        abs_diff = abs(diff)

        # Adaptive noise level estimation
        noise_level = (self.variance_map[cell_id] ** 0.5) + 1e-6

        if abs_diff < self.noise_factor * noise_level:
            # Small variation → update background slowly (learn new normal)
            self.background_map[cell_id] = (1 - self.alpha) * bg + self.alpha * current_power
            self.history[cell_id] = 0

            # Slowly update variance estimate
            self.variance_map[cell_id] = 0.96 * self.variance_map[cell_id] + 0.04 * (diff ** 2)
            return False

        elif diff > self.target_factor * noise_level:
            # Strong positive difference → potential target
            self.history[cell_id] = self.history.get(cell_id, 0) + 1
            
            if self.history[cell_id] >= self.persistence_frames:
                return True  # Persistent target detected
            return False

        else:
            # Difference in gray zone → reset persistence
            self.history[cell_id] = 0
            return False

    def reset_background(self, cell_id=None):
        """Reset background for a specific cell or the entire map."""
        if cell_id is None:
            self.background_map.clear()
            self.variance_map.clear()
            self.history.clear()
        else:
            self.background_map.pop(cell_id, None)
            self.variance_map.pop(cell_id, None)
            self.history.pop(cell_id, None)


# ====================== EXAMPLE USAGE ======================
if __name__ == "__main__":
    # Create instance
    clutter = ClutterLearner(alpha=0.008, noise_factor=1.8, target_factor=3.5, persistence_frames=4)

    test_cell = "az_045_range_1500"
    readings = [0.15, 0.22, 0.18, 0.75, 0.82, 0.79, 0.25, 0.19, 0.88, 0.91, 0.20]

    print("Simulating radar frames for cell", test_cell)
    for i, power in enumerate(readings, 1):
        is_target = clutter.update(test_cell, power)
        status = ">>> TARGET DETECTED <<<" if is_target else "background"
        print(f"Frame {i:2d} | Power: {power:.3f}  →  {status}")

**Comment by JJassonn69** [comment]
Thank you for the idea. @enriqueherbertag-lgtm. Closing this as this can be done any any agent with your specific prompt.
